### PR TITLE
Nested*Depth Rules (#248)

### PR DESCRIPTION
**Pull request by bobisageek**
Add rules for checking the depth of `if`, `for`, `try`, and `when` blocks/expressions. I think the speks are relatively thorough, but I'm certainly open to being convinced otherwise. 

Probably very open for debate is the default thresholds for each rule. I went for thresholds of 2 for `if` and `for`, and 1 for `try` and `when`. My rationale is that `if` and `for` are generally more constrained than the others in the complexity/number of their possible outcomes. For example, `if` is a "this or that" proposition, but `when` has a more arbitrary number of outcomes.

### CODE REVIEW

**Comment by schalkms** [comment]
This is a really nice first-time contribution. I really appreciate this. Big thanks!

Appropriate testing can either be done with this [script](https://github.com/arturbosch/detekt/blob/master/scripts/get_analysis_projects.groovy) or manually over the mentioned projects in this script. These test projects are usually a good indicator how stable a rule is.

**Comment by schalkms** [comment]
@bobisageek I'll happily review your code soon.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+
+	subject { NestedTryDepth(threshold = 2) }
+
+	describe("simple nested tries") {
+
+		it("should detect too many nested tries") {
+			val code = """
+				fun tooDeep() {
+	            	try {
+	            		try {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt]: formatting looks weird here

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+		it("should detect multiple findings in one file") {
+			subject.lint(multipleTryDepthFindingsInOneFile)
+			assertThat(subject.findings).hasSize(2)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 5 }).hasSize(1)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 3 }).hasSize(1)
+		}
+	}
+})
+
+const val triesInOtherBlocks = """
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+					}
+                }
+            } catch (e: Exception) {
+			}
+        }
+    }
+"""
+
+const val multipleTryDepthFindingsInOneFile = """
+    fun wayTooDeep() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+            if (2 > 1) {
+                if (3 > 2) {
+    			}
+    		}
+    	}
+    }
+"""
+
+const val nestedIfCode = """
+	override fun procedure(node: ASTNode) {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+"""
+
+const val multipleFindingsInOneFile = """
+    fun wayTooDeep() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+		it("should not count else if as two") {
+			subject.lint(nestedIfCode)
+			assertThat(subject.findings).isEmpty()
+		}
+	}
+})
+
+const val ifsInOtherBlocks = """
+    fun buriedInBlocks() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+			assertThat(subject.findings).hasSize(2)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 3 }).hasSize(1)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 2 }).hasSize(1)
+		}
+
+	}
+})
+
+const val whensInOtherBlocks = """
+    fun buriedInBlocks() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+	                	}
+					}
+	            }
+	        }
+	    }
+    }
+"""
+
+const val multipleWhenDepthFindingsInOneFile = """
+    fun wayTooDeep() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt [hunk 9]
```diff
@@ -0,0 +1,129 @@
...
+
+class NestedWhenDepthSpec : SubjectSpek<NestedWhenDepth>({
+
+	subject { NestedWhenDepth(threshold = 1) }
+
+	describe("simple nested whens") {
+
+		it("should detect too many nested whens") {
+			val code = """
+				fun tooDeep() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-cli/src/main/resources/default-detekt-config.yml [hunk 10]
```diff
@@ -97,6 +97,18 @@ complexity:
...
+    threshold: 2
+  NestedIfDepth:
+    active: false
+    threshold: 2
+  NestedTryDepth:
+    active: false
+    threshold: 1
+  NestedWhenDepth:
+    active: false
+    threshold: 1
```
Comment (schalkms) [inline detekt-cli/src/main/resources/default-detekt-config.yml]: I like these default values.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepth.kt [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+import io.gitlab.arturbosch.detekt.api.ThresholdedCodeSmell
+import io.gitlab.arturbosch.detekt.rules.collectByType
+import org.jetbrains.kotlin.psi.KtForExpression
+
+/**
+ * This rule reports excessive nesting depth of `for` expressions. Excessively nested for loops can be
+ * harder to manage in one's head, and commonly masks the intended behavior of the function.
+ *
+ * Prefer extracting the nested code into well-named functions and/or using expressive comprehensions to
+ * make it easier to understand.
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepth.kt]: Good comment!

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepth.kt [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+class NestedForDepth(config: Config = Config.empty,
+					 threshold: Int = DEFAULT_ACCEPTED_NESTING) : ThresholdRule(config, threshold) {
+
+	override val issue = Issue("NestedForDepth",
+			Severity.Maintainability,
+			"Excessive nesting leads to hidden complexity and obscures meaning. " +
+					"Prefer extracting code to make it easier to understand.",
+			Debt.TWENTY_MINS)
+
+	private val seenFors: MutableSet<KtForExpression> = mutableSetOf()
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepth.kt]: It think it makes more sense to move this mutable set into a private class like in `NestedBlockDepth` and creating an object of it while visiting a named function.
This has the advantage that there is no shared mutable state.

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+class NestedIfDepth(config: Config = Config.empty,
+					threshold: Int = DEFAULT_ACCEPTED_NESTING) : ThresholdRule(config, threshold) {
+
+	override val issue = Issue("NestedIfDepth",
+			Severity.Maintainability,
+			"Excessive nesting leads to hidden complexity and obscures meaning. " +
+					"Prefer extracting code to make it easier to understand.",
+			Debt.TWENTY_MINS)
+
+	private val seenIfs: MutableSet<KtIfExpression> = mutableSetOf()
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt]: Same here regarding shared mutable state

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+import io.gitlab.arturbosch.detekt.rules.collectByType
+import org.jetbrains.kotlin.psi.KtContainerNodeForControlStructureBody
+import org.jetbrains.kotlin.psi.KtIfExpression
+
+/**
+ * This rule reports excessive nesting depth of `if` expressions. Excessively nested `if`s increase the complexity of
+ * a function by increasing the number of possible execution paths and state dependencies.
+ *
+ * Prefer simplifying the conditional logic or extracting the nested code into well-named functions
+ * to make it easier to understand.
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt]: I really like the consistency in the rule description. Good stuff!

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+		if (nestedIfs.size > threshold) {
+			report(ThresholdedCodeSmell(issue,
+					Entity.from(expression),
+					Metric("NESTING DEPTH", nestedIfs.size, threshold),
+					"These if expressions are nested too deeply."))
+		}
+	}
+
+	companion object {
+
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepth.kt]: Can we please remove the unnecessary blank line?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepth.kt [hunk 16]
```diff
@@ -0,0 +1,58 @@
...
+		if (nestedTries.size > threshold) {
+			report(ThresholdedCodeSmell(issue,
+					Entity.from(expression),
+					Metric("NESTING DEPTH", nestedTries.size, threshold),
+					"These try blocks are nested too deeply."))
+		}
+	}
+
+	companion object {
+
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepth.kt]: Can we please remove the unnecessary blank line?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepth.kt [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+		if (nestedWhens.size > threshold) {
+			report(ThresholdedCodeSmell(issue,
+					Entity.from(expression),
+					Metric("NESTING DEPTH", nestedWhens.size, threshold),
+					"These when expressions are nested too deeply."))
+		}
+	}
+
+	companion object {
+
```
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepth.kt]: Can we please remove the unnecessary blank line?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+                        }
+                    }
+                }
+            }
+        }
+    }
+"""
+
+const val multipleForFindingsInOneFile = """
+    fun wayTooDeep() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+			subject.lint(multipleForFindingsInOneFile)
+			assertThat(subject.findings).hasSize(2)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 5 }).hasSize(1)
+			assertThat(subject.findings.filter { it.metricByType("NESTING DEPTH")?.value == 3 }).hasSize(1)
+		}
+	}
+})
+
+const val forsInOtherBlocks = """
+    fun buriedInBlocks() {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt]: Can we please put this in a separate file to be more maintainable?

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+	describe("simple nested fors") {
+
+		it("should detect too many nested fors") {
+			val code = """
+				fun tooDeep() {
+					for (i in 1..3) {
+						for (j in 1..3) {
+							for (k in 1..3) {
+								for (l in 1..3) {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt]: The last `for` statement can be omitted. It should report the third loop anyways.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt [hunk 21]
```diff
@@ -0,0 +1,105 @@
...
+
+import io.gitlab.arturbosch.detekt.test.lint
+import org.assertj.core.api.Assertions.assertThat
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.subject.SubjectSpek
+
+class NestedForDepthSpec : SubjectSpek<NestedForDepth>({
+
+	subject { NestedForDepth(threshold = 2) }
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt]: I would suggest setting the threshold to `1` in order to write less test code.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt (hunk 18 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedForDepthSpec.kt]: Why do you test all this stuff twice? The tests in the first describe block already cover these cases.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+
+	describe("simple nested ifs") {
+
+		it("should detect too many nested ifs") {
+			val code = """
+				fun tooDeep() {
+					if (1 > 0) {
+						if (2 > 1) {
+							if (3 > 2) {
+								if (4 > 3) {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: The last `if` statement can be omitted. It should report the third `if` anyways.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+
+import io.gitlab.arturbosch.detekt.test.lint
+import org.assertj.core.api.Assertions.assertThat
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.subject.SubjectSpek
+
+class NestedIfDepthSpec : SubjectSpek<NestedIfDepth>({
+
+	subject { NestedIfDepth(threshold = 2) }
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: I would suggest setting the threshold to `1` in order to write less test code.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt (hunk 5 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: Why do you test all this stuff twice? The tests in the first describe block already cover these cases.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt (hunk 4 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedIfDepthSpec.kt]: I like it. This is thoroughly tested.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt [hunk 24]
```diff
@@ -0,0 +1,123 @@
...
+
+import io.gitlab.arturbosch.detekt.test.lint
+import org.assertj.core.api.Assertions.assertThat
+import org.jetbrains.spek.api.dsl.describe
+import org.jetbrains.spek.api.dsl.it
+import org.jetbrains.spek.subject.SubjectSpek
+
+class NestedTryDepthSpec : SubjectSpek<NestedTryDepth>({
+
+	subject { NestedTryDepth(threshold = 2) }
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt]: I would suggest setting the threshold to `1` in order to write less test code.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt (hunk 3 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedTryDepthSpec.kt]: Why do you test all this stuff twice? The tests in the first describe block already cover these cases.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt [hunk 25]
```diff
@@ -0,0 +1,129 @@
...
+				    when {
+				        1 > 0 -> {
+				            when {
+				                2 > 1 -> {
+				                    when {
+				                        3 > 2 -> {
```
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt]: The last two `when` statements can be omitted. It should report the second `when` anyways.

File: detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt (hunk 8 above)
Comment (schalkms) [inline detekt-rules/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/NestedWhenDepthSpec.kt]: This stuff should be covered by the test code in `whensInOtherBlocks` and can be removed

**Comment by schalkms** [comment]
@bobisageek any update on this PR? Do you need assistance?

**Comment by bobisageek** [comment]
No updates, I just had the holidays and then work has sort of picked up, so I've had to shelve the side projects for a little while. Since I'm sure there will be a bunch of updates and large re-writes for the shared mutable state, I'll just close this and open a clean new one if/when I can get to the rewrite.

**Pull request closed without merging**